cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java

```
@@ -201,22 +207,62 @@ public void configure(Map<String, ?> configs) {
    *
    * @return  MetricsRegistry with Kafka metrics
    */
   private static MetricsRegistry metricsRegistry() {
+    Object metricsRegistry;
+    Class<?> metrics;
+
     try {
-      Class.forName("kafka.metrics.KafkaYammerMetrics");
-      LOG.info("KafkaYammerMetrics found and will be used.");
-      return KafkaYammerMetrics.defaultRegistry();
+      // First we try to get the KafkaYammerMetrics class for Kafka 3.3+
+      metrics = Class.forName(YAMMER_METRICS_IN_KAFKA_3_3_AND_LATER);
+      LOG.info("Found class {} for Kafka 3.3 and newer.", YAMMER_METRICS_IN_KAFKA_3_3_AND_LATER);
     } catch (ClassNotFoundException e) {
-      LOG.info("KafkaYammerMetrics not found. Metrics will be used.");
-      return Metrics.defaultRegistry();
+      LOG.info("Class {} not found. We are probably on Kafka 3.2 or older.", YAMMER_METRICS_IN_KAFKA_3_3_AND_LATER);
+
+      // We did not find the KafkaYammerMetrics class from Kafka 3.3+. So we are probably on older Kafka version
+      //     => we will try the older class for Kafka 2.6+.
+      try {
+        metrics = Class.forName(YAMMER_METRICS_IN_KAFKA_2_6_AND_LATER);
+        LOG.info("Found class {} for Kafka 2.6 and newer.", YAMMER_METRICS_IN_KAFKA_2_6_AND_LATER);
+      } catch (ClassNotFoundException ee) {
+        LOG.info("Class {} not found. We are probably on Kafka 2.5 or older.", YAMMER_METRICS_IN_KAFKA_2_6_AND_LATER);
+
+        // We did not find the KafkaYammerMetrics class from Kafka 2.6+. So we are probably on older Kafka version
+        //     => we will try the older class for Kafka 2.5-.
+        try {
+          metrics = Class.forName(YAMMER_METRICS_IN_KAFKA_2_5_AND_EARLIER);
+          LOG.info("Found class {} for Kafka 2.5 and earlier.", YAMMER_METRICS_IN_KAFKA_2_5_AND_EARLIER);
+        } catch (ClassNotFoundException eee) {
+          // No class was found for any Kafka version => we should fail
+          throw new RuntimeException("Failed to find Yammer Metrics class", eee);
```

Is there a reason that we throws exception instead of returning Metrics.defaultRegistry()？ I'm asking because I see the code was returning Metrics.defaultRegistry() before.

---

We do return Metrics.defaultRegistry() by the end of the method if the class and method exists, this makes it consistent with how we would check the defaultRegistry() methods of the other classes

---

Ah I see. Thanks for the explanation.